File: ulam/ulam.py
# w/  numba:  5,856,608,054ns used to find 7,584 Ulam numbers <= 100,000.
# w/o numba: 15,245,441,791ns used to find 7,584 Ulam numbers <= 100,000.
import csv
import time
from typing import Callable

from absl import app
from absl import flags
import numba

FLAGS = flags.FLAGS
flags.DEFINE_integer('limit', 100000, '', lower_bound = 1000)
flags.DEFINE_string('output_csv', 'ulam_100k_py.csv', '')

@numba.njit('boolean(int64, ListType(int64))', nogil=True)
def is_ulam_number(num: int, ulam_numbers: list[int]) -> bool:
    i = 0
    j = len(ulam_numbers) - 1
    fnd = False
    while i < j:
        total = ulam_numbers[i] + ulam_numbers[j]
        if total == num:
            if fnd:
                return False
            fnd = True
            i += 1
            j -= 1
        elif total < num:
            i += 1
        else:
            j -= 1
    return fnd

def main(argv: list[str]):
    ulam_numbers = numba.typed.List()
    ulam_numbers.append(1)
    ulam_numbers.append(2)
    elapsed_ns_list = [0, 0]
    total_elapsed_ns = 0
    start_ns = time.time_ns()
    for i in range(3, FLAGS.limit+1):
        if is_ulam_number(i, ulam_numbers):
            elapsed_ns = time.time_ns() - start_ns
            print(f'{elapsed_ns:,} ns used to find Ulam number {i:,}.')
            total_elapsed_ns += elapsed_ns
            start_ns = time.time_ns()
            ulam_numbers.append(i)
            elapsed_ns_list.append(elapsed_ns)
    print(f'{total_elapsed_ns:,}ns used to find {len(ulam_numbers):,} Ulam numbers <= {FLAGS.limit:,}.')
    with open(FLAGS.output_csv, 'w', newline='') as csv_file:
        csv_writer = csv.writer(csv_file)
        csv_writer.writerow(['ulam_num', 'elapsed_ns'])
        for i in range(len(ulam_numbers)):
            csv_writer.writerow([ulam_numbers[i], elapsed_ns_list[i]])

app.run(main)

File: ulam/ulam.cc
// 406518412 ns used to find 7584 Ulam numbers.

#include <inttypes.h>

#include <chrono>
#include <fstream>
#include <iostream>
#include <numeric>
#include <vector>

#include <gflags/gflags.h>

DEFINE_int64(limit, 100000ll, "");
DEFINE_string(output_csv, "ulam_100k_cc.csv", "");

namespace {

bool is_ulam_number(int64_t num, const std::vector<int64_t>& ulam_numbers) {
    int i = 0, j = ulam_numbers.size() - 1;
    bool fnd = false;
    while (i < j) {
        const int64_t sum = ulam_numbers[i] + ulam_numbers[j];
        if (sum == num) {
            if (fnd) return false;
            fnd = true;
            i++;
            j--;
        } else if (sum < num) {
            i++;
        } else {
            j--;
        }
    }
    return fnd;
}

}  // namespace


int main(int argc, char* argv[]) {
    gflags::ParseCommandLineFlags(&argc, &argv, true);

    std::vector<int64_t> ulam_numbers;
    std::vector<int64_t> elapsed_ns_list;
    ulam_numbers.push_back(1);
    ulam_numbers.push_back(2);
    elapsed_ns_list.push_back(0);
    elapsed_ns_list.push_back(0);
    auto start = std::chrono::high_resolution_clock::now();
    int64_t total_elapsed_ns = 0;
    for (int i = 3; i <= FLAGS_limit; i++) {
        if (is_ulam_number(i, ulam_numbers)) {
            const std::chrono::duration<double> elapsed =
                std::chrono::high_resolution_clock::now() - start;
            const int64_t elapsed_ns =
                std::chrono::duration_cast<std::chrono::nanoseconds>(elapsed).count();
            std::cout << elapsed_ns << " ns used to find Ulam number " << i << std::endl;
            total_elapsed_ns += elapsed_ns;
            start = std::chrono::high_resolution_clock::now();
            ulam_numbers.push_back(i);
            elapsed_ns_list.push_back(elapsed_ns);
        }
    }
    const std::chrono::duration<double> elapsed =
        std::chrono::high_resolution_clock::now() - start;
    total_elapsed_ns +=
        std::chrono::duration_cast<std::chrono::nanoseconds>(elapsed).count();
    std::cout << total_elapsed_ns << " ns used to find " << ulam_numbers.size() << " Ulam numbers.\n";

    std::ofstream csv_file;
    csv_file.open(FLAGS_output_csv);
    csv_file << "ulam_num,elapsed_ns\n";
    for (int i = 0; i < ulam_numbers.size(); i++) {
        csv_file << ulam_numbers[i] << "," << elapsed_ns_list[i] << std::endl;
    }
    csv_file.close();
}
